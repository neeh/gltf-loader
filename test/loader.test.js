const path = require('node:path');
const compiler = require('./compiler.js');

async function readOutputFile(stats, filePath) {
  const outputFs = stats.compilation.compiler.outputFileSystem;
  const outputFilePath = path.join(stats.compilation.options.output.path, filePath);

  return outputFs.promises.readFile(outputFilePath, 'utf8');
}

test('Outputs valid .gltf', async () => {
  const stats = await compiler('./fixtures/cube.js', {
    options: {
      name: '[name].[hash:8].[ext]'
    }
  });
  expect(Object.keys(stats.compilation.assets).sort()).toMatchSnapshot();

  const cubeSource = await readOutputFile(stats, 'cube.6a77dc81.gltf');
  expect(cubeSource).toMatchSnapshot();
});

test('Throws an error on invalid .gltf', async () => {
  expect.assertions(2);
  await compiler('./fixtures/cube-invalid.js', {
    options: {
      name: '[name].[hash:8].[ext]'
    }
  }).catch(errors => {
    expect(errors).toHaveLength(1);
    expect(errors[0].message.indexOf('Error: Invalid glTF file')).not.toBe(-1);
  });
});

test('Shares assets over multiple .gltf', async () => {
  const stats = await compiler('./fixtures/shared.js', {
    options: {
      name: '[name].[hash:8].[ext]'
    }
  });
  expect(Object.keys(stats.compilation.assets).sort()).toMatchSnapshot();

  // Use "assets/cube.c86430a0.png"
  const cubeSource = await readOutputFile(stats, 'cube.6a77dc81.gltf');
  expect(cubeSource).toMatchSnapshot();

  // Also uses "assets/cube.c86430a0.png"
  const facesSource = await readOutputFile(stats, 'faces.80de93ce.gltf');
  expect(facesSource).toMatchSnapshot();
});

test('Outputs relative paths in the exported .gltf', async () => {
  const stats = await compiler('./fixtures/cube.js', {
    options: {
      name: 'assets/[name].[hash:8].[ext]'
    }
  });
  expect(Object.keys(stats.compilation.assets).sort()).toMatchSnapshot();

  // Image URI is "cube.c86430a0.png" (same folder)
  const cubeSource = await readOutputFile(stats, 'assets/cube.6a77dc81.gltf');
  expect(cubeSource).toMatchSnapshot();
});

test('Outputs relative paths in the exported .gltf when a public path is used', async () => {
  const stats = await compiler('./fixtures/cube.js', {
    publicPath: 'https://website.com/',
    options: {
      name: 'shared/[name].[hash:8].[ext]'
    }
  });
  expect(Object.keys(stats.compilation.assets).sort()).toMatchSnapshot();

  // Image URI should be the relative path "../assets/cube.c86430a0.png"
  const cubeSource = await readOutputFile(stats, 'shared/cube.6a77dc81.gltf');
  expect(cubeSource).toMatchSnapshot();
});

test('Outputs an inline .gltf inside the module when the option is used', async () => {
  const stats = await compiler('./fixtures/cube.js', {
    options: {
      inline: true,
      name: 'shared/[name].[hash:8].[ext]'
    }
  });
  expect(Object.keys(stats.compilation.assets).sort()).toMatchSnapshot();

  const { modules } = stats.toJson({ source: true });
  const outputModule = modules.find(module => module.name === './fixtures/assets/cube.gltf');

  // .gltf should be emitted in the module itself
  expect(outputModule.source).toMatchSnapshot();
});
