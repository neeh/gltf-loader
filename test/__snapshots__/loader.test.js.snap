// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outputs an inline .gltf inside the module when the option is used 1`] = `
[
  "assets/cube.c86430a0.png",
  "assets/cube.d0ee63e1.bin",
  "main.js",
]
`;

exports[`Outputs an inline .gltf inside the module when the option is used 2`] = `
"import "!!../../../src/asset-loader.js!./cube.bin";
import "!!../../../src/asset-loader.js!./cube.png";
export default {"asset":{"generator":"Khronos glTF Blender I/O v4.2.69","version":"2.0"},"scene":0,"scenes":[{"name":"Scene","nodes":[0]}],"nodes":[{"mesh":0,"name":"Cube"}],"materials":[{"doubleSided":true,"name":"Cube","pbrMetallicRoughness":{"baseColorTexture":{"index":0},"metallicFactor":0,"roughnessFactor":0.5}}],"meshes":[{"name":"Cube.001","primitives":[{"attributes":{"POSITION":0,"NORMAL":1,"TEXCOORD_0":2},"indices":3,"material":0}]}],"textures":[{"sampler":0,"source":0}],"images":[{"mimeType":"image/png","name":"cube","uri":"assets/cube.c86430a0.png"}],"accessors":[{"bufferView":0,"componentType":5126,"count":24,"max":[1,1,1],"min":[-1,-1,-1],"type":"VEC3"},{"bufferView":1,"componentType":5126,"count":24,"type":"VEC3"},{"bufferView":2,"componentType":5126,"count":24,"type":"VEC2"},{"bufferView":3,"componentType":5123,"count":36,"type":"SCALAR"}],"bufferViews":[{"buffer":0,"byteLength":288,"byteOffset":0,"target":34962},{"buffer":0,"byteLength":288,"byteOffset":288,"target":34962},{"buffer":0,"byteLength":192,"byteOffset":576,"target":34962},{"buffer":0,"byteLength":72,"byteOffset":768,"target":34963}],"samplers":[{"magFilter":9729,"minFilter":9987}],"buffers":[{"byteLength":840,"uri":"assets/cube.d0ee63e1.bin"}]}"
`;

exports[`Outputs relative paths in the exported .gltf 1`] = `
[
  "assets/cube.6a77dc81.gltf",
  "assets/cube.c86430a0.png",
  "assets/cube.d0ee63e1.bin",
  "main.js",
]
`;

exports[`Outputs relative paths in the exported .gltf 2`] = `"{"asset":{"generator":"Khronos glTF Blender I/O v4.2.69","version":"2.0"},"scene":0,"scenes":[{"name":"Scene","nodes":[0]}],"nodes":[{"mesh":0,"name":"Cube"}],"materials":[{"doubleSided":true,"name":"Cube","pbrMetallicRoughness":{"baseColorTexture":{"index":0},"metallicFactor":0,"roughnessFactor":0.5}}],"meshes":[{"name":"Cube.001","primitives":[{"attributes":{"POSITION":0,"NORMAL":1,"TEXCOORD_0":2},"indices":3,"material":0}]}],"textures":[{"sampler":0,"source":0}],"images":[{"mimeType":"image/png","name":"cube","uri":"cube.c86430a0.png"}],"accessors":[{"bufferView":0,"componentType":5126,"count":24,"max":[1,1,1],"min":[-1,-1,-1],"type":"VEC3"},{"bufferView":1,"componentType":5126,"count":24,"type":"VEC3"},{"bufferView":2,"componentType":5126,"count":24,"type":"VEC2"},{"bufferView":3,"componentType":5123,"count":36,"type":"SCALAR"}],"bufferViews":[{"buffer":0,"byteLength":288,"byteOffset":0,"target":34962},{"buffer":0,"byteLength":288,"byteOffset":288,"target":34962},{"buffer":0,"byteLength":192,"byteOffset":576,"target":34962},{"buffer":0,"byteLength":72,"byteOffset":768,"target":34963}],"samplers":[{"magFilter":9729,"minFilter":9987}],"buffers":[{"byteLength":840,"uri":"cube.d0ee63e1.bin"}]}"`;

exports[`Outputs relative paths in the exported .gltf when a public path is used 1`] = `
[
  "assets/cube.c86430a0.png",
  "assets/cube.d0ee63e1.bin",
  "main.js",
  "shared/cube.6a77dc81.gltf",
]
`;

exports[`Outputs relative paths in the exported .gltf when a public path is used 2`] = `"{"asset":{"generator":"Khronos glTF Blender I/O v4.2.69","version":"2.0"},"scene":0,"scenes":[{"name":"Scene","nodes":[0]}],"nodes":[{"mesh":0,"name":"Cube"}],"materials":[{"doubleSided":true,"name":"Cube","pbrMetallicRoughness":{"baseColorTexture":{"index":0},"metallicFactor":0,"roughnessFactor":0.5}}],"meshes":[{"name":"Cube.001","primitives":[{"attributes":{"POSITION":0,"NORMAL":1,"TEXCOORD_0":2},"indices":3,"material":0}]}],"textures":[{"sampler":0,"source":0}],"images":[{"mimeType":"image/png","name":"cube","uri":"../assets/cube.c86430a0.png"}],"accessors":[{"bufferView":0,"componentType":5126,"count":24,"max":[1,1,1],"min":[-1,-1,-1],"type":"VEC3"},{"bufferView":1,"componentType":5126,"count":24,"type":"VEC3"},{"bufferView":2,"componentType":5126,"count":24,"type":"VEC2"},{"bufferView":3,"componentType":5123,"count":36,"type":"SCALAR"}],"bufferViews":[{"buffer":0,"byteLength":288,"byteOffset":0,"target":34962},{"buffer":0,"byteLength":288,"byteOffset":288,"target":34962},{"buffer":0,"byteLength":192,"byteOffset":576,"target":34962},{"buffer":0,"byteLength":72,"byteOffset":768,"target":34963}],"samplers":[{"magFilter":9729,"minFilter":9987}],"buffers":[{"byteLength":840,"uri":"../assets/cube.d0ee63e1.bin"}]}"`;

exports[`Outputs valid .gltf 1`] = `
[
  "assets/cube.c86430a0.png",
  "assets/cube.d0ee63e1.bin",
  "cube.6a77dc81.gltf",
  "main.js",
]
`;

exports[`Outputs valid .gltf 2`] = `"{"asset":{"generator":"Khronos glTF Blender I/O v4.2.69","version":"2.0"},"scene":0,"scenes":[{"name":"Scene","nodes":[0]}],"nodes":[{"mesh":0,"name":"Cube"}],"materials":[{"doubleSided":true,"name":"Cube","pbrMetallicRoughness":{"baseColorTexture":{"index":0},"metallicFactor":0,"roughnessFactor":0.5}}],"meshes":[{"name":"Cube.001","primitives":[{"attributes":{"POSITION":0,"NORMAL":1,"TEXCOORD_0":2},"indices":3,"material":0}]}],"textures":[{"sampler":0,"source":0}],"images":[{"mimeType":"image/png","name":"cube","uri":"assets/cube.c86430a0.png"}],"accessors":[{"bufferView":0,"componentType":5126,"count":24,"max":[1,1,1],"min":[-1,-1,-1],"type":"VEC3"},{"bufferView":1,"componentType":5126,"count":24,"type":"VEC3"},{"bufferView":2,"componentType":5126,"count":24,"type":"VEC2"},{"bufferView":3,"componentType":5123,"count":36,"type":"SCALAR"}],"bufferViews":[{"buffer":0,"byteLength":288,"byteOffset":0,"target":34962},{"buffer":0,"byteLength":288,"byteOffset":288,"target":34962},{"buffer":0,"byteLength":192,"byteOffset":576,"target":34962},{"buffer":0,"byteLength":72,"byteOffset":768,"target":34963}],"samplers":[{"magFilter":9729,"minFilter":9987}],"buffers":[{"byteLength":840,"uri":"assets/cube.d0ee63e1.bin"}]}"`;

exports[`Shares assets over multiple .gltf 1`] = `
[
  "assets/cube.c86430a0.png",
  "assets/cube.d0ee63e1.bin",
  "assets/faces.cff79296.bin",
  "cube.6a77dc81.gltf",
  "faces.80de93ce.gltf",
  "main.js",
]
`;

exports[`Shares assets over multiple .gltf 2`] = `"{"asset":{"generator":"Khronos glTF Blender I/O v4.2.69","version":"2.0"},"scene":0,"scenes":[{"name":"Scene","nodes":[0]}],"nodes":[{"mesh":0,"name":"Cube"}],"materials":[{"doubleSided":true,"name":"Cube","pbrMetallicRoughness":{"baseColorTexture":{"index":0},"metallicFactor":0,"roughnessFactor":0.5}}],"meshes":[{"name":"Cube.001","primitives":[{"attributes":{"POSITION":0,"NORMAL":1,"TEXCOORD_0":2},"indices":3,"material":0}]}],"textures":[{"sampler":0,"source":0}],"images":[{"mimeType":"image/png","name":"cube","uri":"assets/cube.c86430a0.png"}],"accessors":[{"bufferView":0,"componentType":5126,"count":24,"max":[1,1,1],"min":[-1,-1,-1],"type":"VEC3"},{"bufferView":1,"componentType":5126,"count":24,"type":"VEC3"},{"bufferView":2,"componentType":5126,"count":24,"type":"VEC2"},{"bufferView":3,"componentType":5123,"count":36,"type":"SCALAR"}],"bufferViews":[{"buffer":0,"byteLength":288,"byteOffset":0,"target":34962},{"buffer":0,"byteLength":288,"byteOffset":288,"target":34962},{"buffer":0,"byteLength":192,"byteOffset":576,"target":34962},{"buffer":0,"byteLength":72,"byteOffset":768,"target":34963}],"samplers":[{"magFilter":9729,"minFilter":9987}],"buffers":[{"byteLength":840,"uri":"assets/cube.d0ee63e1.bin"}]}"`;

exports[`Shares assets over multiple .gltf 3`] = `"{"asset":{"generator":"Khronos glTF Blender I/O v4.2.69","version":"2.0"},"scene":0,"scenes":[{"name":"Scene","nodes":[0,1,2,3,4,5]}],"nodes":[{"mesh":0,"name":"Face0","translation":[-1.5,0,-0.75]},{"mesh":1,"name":"Face1","translation":[0,0,-0.75]},{"mesh":2,"name":"Face2","translation":[1.5,0,-0.75]},{"mesh":3,"name":"Face3","translation":[-1.5,0,0.75]},{"mesh":4,"name":"Face4","translation":[0,0,0.75]},{"mesh":5,"name":"Face5","translation":[1.5,0,0.75]}],"materials":[{"doubleSided":true,"name":"Face","pbrMetallicRoughness":{"baseColorTexture":{"index":0},"metallicFactor":0,"roughnessFactor":0.5}}],"meshes":[{"name":"Face0","primitives":[{"attributes":{"POSITION":0,"NORMAL":1,"TEXCOORD_0":2},"indices":3,"material":0}]},{"name":"Face1","primitives":[{"attributes":{"POSITION":4,"NORMAL":5,"TEXCOORD_0":6},"indices":3,"material":0}]},{"name":"Face2","primitives":[{"attributes":{"POSITION":7,"NORMAL":8,"TEXCOORD_0":9},"indices":3,"material":0}]},{"name":"Face3","primitives":[{"attributes":{"POSITION":10,"NORMAL":11,"TEXCOORD_0":12},"indices":3,"material":0}]},{"name":"Face4","primitives":[{"attributes":{"POSITION":13,"NORMAL":14,"TEXCOORD_0":15},"indices":3,"material":0}]},{"name":"Face5","primitives":[{"attributes":{"POSITION":16,"NORMAL":17,"TEXCOORD_0":18},"indices":3,"material":0}]}],"textures":[{"sampler":0,"source":0}],"images":[{"mimeType":"image/png","name":"cube","uri":"assets/cube.c86430a0.png"}],"accessors":[{"bufferView":0,"componentType":5126,"count":4,"max":[0.5,0,0.5],"min":[-0.5,0,-0.5],"type":"VEC3"},{"bufferView":1,"componentType":5126,"count":4,"type":"VEC3"},{"bufferView":2,"componentType":5126,"count":4,"type":"VEC2"},{"bufferView":3,"componentType":5123,"count":6,"type":"SCALAR"},{"bufferView":4,"componentType":5126,"count":4,"max":[0.5,0,0.5],"min":[-0.5,0,-0.5],"type":"VEC3"},{"bufferView":5,"componentType":5126,"count":4,"type":"VEC3"},{"bufferView":6,"componentType":5126,"count":4,"type":"VEC2"},{"bufferView":7,"componentType":5126,"count":4,"max":[0.5,0,0.5],"min":[-0.5,0,-0.5],"type":"VEC3"},{"bufferView":8,"componentType":5126,"count":4,"type":"VEC3"},{"bufferView":9,"componentType":5126,"count":4,"type":"VEC2"},{"bufferView":10,"componentType":5126,"count":4,"max":[0.5,0,0.5],"min":[-0.5,0,-0.5],"type":"VEC3"},{"bufferView":11,"componentType":5126,"count":4,"type":"VEC3"},{"bufferView":12,"componentType":5126,"count":4,"type":"VEC2"},{"bufferView":13,"componentType":5126,"count":4,"max":[0.5,0,0.5],"min":[-0.5,0,-0.5],"type":"VEC3"},{"bufferView":14,"componentType":5126,"count":4,"type":"VEC3"},{"bufferView":15,"componentType":5126,"count":4,"type":"VEC2"},{"bufferView":16,"componentType":5126,"count":4,"max":[0.5,0,0.5],"min":[-0.5,0,-0.5],"type":"VEC3"},{"bufferView":17,"componentType":5126,"count":4,"type":"VEC3"},{"bufferView":18,"componentType":5126,"count":4,"type":"VEC2"}],"bufferViews":[{"buffer":0,"byteLength":48,"byteOffset":0,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":48,"target":34962},{"buffer":0,"byteLength":32,"byteOffset":96,"target":34962},{"buffer":0,"byteLength":12,"byteOffset":128,"target":34963},{"buffer":0,"byteLength":48,"byteOffset":140,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":188,"target":34962},{"buffer":0,"byteLength":32,"byteOffset":236,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":268,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":316,"target":34962},{"buffer":0,"byteLength":32,"byteOffset":364,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":396,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":444,"target":34962},{"buffer":0,"byteLength":32,"byteOffset":492,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":524,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":572,"target":34962},{"buffer":0,"byteLength":32,"byteOffset":620,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":652,"target":34962},{"buffer":0,"byteLength":48,"byteOffset":700,"target":34962},{"buffer":0,"byteLength":32,"byteOffset":748,"target":34962}],"samplers":[{"magFilter":9729,"minFilter":9987}],"buffers":[{"byteLength":780,"uri":"assets/faces.cff79296.bin"}]}"`;
